File: web_client/spatial_audio.js
// Shared Web Audio spatialization policy for every server-driven source.
// Server coordinates are listener-relative: +X right, +Y forward, +Z up.

export const MAX_AUDIO_POSITION = 1000;
const ATTENUATION_PRECISION_SCALE = 1_000_000;
export const MAX_AUDIO_DISTANCE = Math.ceil(
  Math.sqrt(3) * MAX_AUDIO_POSITION * ATTENUATION_PRECISION_SCALE,
) / ATTENUATION_PRECISION_SCALE;
export const MAX_AUDIO_ROLLOFF = 16;
export const MAX_AUDIO_AUTOMATION_MS = 3_600_000;
export const TABLE_RADIUS = 2;
const ATTENUATION_MODELS = new Set(["none", "linear", "inverse", "exponential"]);
const MOTION_EASINGS = new Set(["linear", "ease-in", "ease-out", "ease-in-out"]);

export function normalizeAudioPosition(value) {
  if (value === undefined || value === null) {
    return undefined;
  }
  if (!Array.isArray(value) || value.length !== 3) {
    return null;
  }
  if (value.some(
    (coordinate) => !Number.isFinite(coordinate)
      || Math.abs(coordinate) > MAX_AUDIO_POSITION,
  )) {
    return null;
  }
  return Object.freeze([...value]);
}

export function panFromPosition(position) {
  const [x, y] = position;
  const horizontalDistance = Math.hypot(x, y);
  return horizontalDistance < 1e-6
    ? 0
    : Math.max(-1, Math.min(1, x / horizontalDistance));
}

export function normalizeDistanceAttenuation(value) {
  if (value === undefined || value === null) {
    return undefined;
  }
  if (typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  const model = value.model;
  if (!ATTENUATION_MODELS.has(model)) {
    return null;
  }
  const keys = Object.keys(value).sort();
  const expected = model === "none"
    ? ["model"]
    : [
      "max_distance",
      "max_gain",
      "min_gain",
      "model",
      "reference_distance",
      "rolloff_factor",
    ];
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    return null;
  }
  if (model === "none") {
    return Object.freeze({ model });
  }
  const {
    reference_distance: referenceDistance,
    max_distance: maxDistance,
    rolloff_factor: rolloffFactor,
    min_gain: minGain,
    max_gain: maxGain,
  } = value;
  if (![referenceDistance, maxDistance, rolloffFactor, minGain, maxGain].every(
    (item) => typeof item === "number" && Number.isFinite(item),
  )) {
    return null;
  }
  const maximumRolloff = model === "linear" ? 1 : MAX_AUDIO_ROLLOFF;
  if (
    !(referenceDistance > 0 && referenceDistance < maxDistance)
    || maxDistance > MAX_AUDIO_DISTANCE
    || rolloffFactor <= 0
    || rolloffFactor > maximumRolloff
    || minGain < 0
    || minGain > maxGain
    || maxGain > 1
  ) {
    return null;
  }
  return Object.freeze({
    model,
    reference_distance: referenceDistance,
    max_distance: maxDistance,
    rolloff_factor: rolloffFactor,
    min_gain: minGain,
    max_gain: maxGain,
  });
}

export function distanceAttenuationGain(position, attenuation) {
  const normalized = normalizeDistanceAttenuation(attenuation);
  if (normalized === null) {
    throw new TypeError("Invalid audio attenuation");
  }
  if (normalized === undefined || normalized.model === "none") {
    return 1;
  }
  if (!position) {
    throw new TypeError("Audio attenuation requires a spatial position");
  }
  const distance = Math.hypot(...position);
  const clampedDistance = Math.min(
    Math.max(distance, normalized.reference_distance),
    normalized.max_distance,
  );
  let gain;
  if (normalized.model === "linear") {
    gain = 1 - normalized.rolloff_factor * (
      (clampedDistance - normalized.reference_distance)
      / (normalized.max_distance - normalized.reference_distance)
    );
  } else if (normalized.model === "inverse") {
    gain = normalized.reference_distance / (
      normalized.reference_distance
      + normalized.rolloff_factor
      * (clampedDistance - normalized.reference_distance)
    );
  } else {
    gain = (clampedDistance / normalized.reference_distance)
      ** -normalized.rolloff_factor;
  }
  return Math.max(normalized.min_gain, Math.min(normalized.max_gain, gain));
}

export function normalizeAudioMotion(value) {
  if (value === undefined || value === null) {
    return undefined;
  }
  if (typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  const keys = Object.keys(value).sort();
  const expected = [
    "destination_position",
    "duration_ms",
    "easing",
    "elapsed_ms",
    "origin_position",
  ];
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    return null;
  }
  const origin = normalizeAudioPosition(value.origin_position);
  const destination = normalizeAudioPosition(value.destination_position);
  if (!origin || !destination) {
    return null;
  }
  if (
    !Number.isSafeInteger(value.duration_ms)
    || value.duration_ms <= 0
    || value.duration_ms > MAX_AUDIO_AUTOMATION_MS
    || !Number.isSafeInteger(value.elapsed_ms)
    || value.elapsed_ms < 0
    || value.elapsed_ms > value.duration_ms
    || !MOTION_EASINGS.has(value.easing)
  ) {
    return null;
  }
  return Object.freeze({
    origin_position: origin,
    destination_position: destination,
    duration_ms: value.duration_ms,
    elapsed_ms: value.elapsed_ms,
    easing: value.easing,
  });
}

export function normalizeAudioGain(value) {
  return typeof value === "number" && Number.isFinite(value) && value >= 0 && value <= 1
    ? value
    : null;
}

export function normalizeAudioGainAutomation(value) {
  if (value === undefined || value === null) {
    return undefined;
  }
  if (typeof value !== "object" || Array.isArray(value)) {
    return null;
  }
  const keys = Object.keys(value).sort();
  const expected = [
    "destination_gain",
    "duration_ms",
    "easing",
    "elapsed_ms",
    "origin_gain",
  ];
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    return null;
  }
  const origin = normalizeAudioGain(value.origin_gain);
  const destination = normalizeAudioGain(value.destination_gain);
  if (
    origin === null
    || destination === null
    || !Number.isSafeInteger(value.duration_ms)
    || value.duration_ms <= 0
    || value.duration_ms > MAX_AUDIO_AUTOMATION_MS
    || !Number.isSafeInteger(value.elapsed_ms)
    || value.elapsed_ms < 0
    || value.elapsed_ms > value.duration_ms
    || !MOTION_EASINGS.has(value.easing)
  ) {
    return null;
  }
  return Object.freeze({
    origin_gain: origin,
    destination_gain: destination,
    duration_ms: value.duration_ms,
    elapsed_ms: value.elapsed_ms,
    easing: value.easing,
  });
}

export function audioMotionProgress(progress, easing) {
  const bounded = Math.max(0, Math.min(1, progress));
  if (easing === "linear") return bounded;
  if (easing === "ease-in") return bounded * bounded;
  if (easing === "ease-out") return 1 - ((1 - bounded) ** 2);
  if (easing === "ease-in-out") {
    return bounded < 0.5
      ? 2 * bounded * bounded
      : 1 - (((-2 * bounded + 2) ** 2) / 2);
  }
  throw new TypeError("Unknown audio motion easing");
}

export function audioMotionPosition(motion, elapsedMs = undefined) {
  const normalized = normalizeAudioMotion(motion);
  if (!normalized) {
    throw new TypeError("Invalid audio motion");
  }
  const elapsed = elapsedMs ?? normalized.elapsed_ms;
  const ratio = audioMotionProgress(
    elapsed / normalized.duration_ms,
    normalized.easing,
  );
  return Object.freeze(normalized.origin_position.map(
    (origin, index) => origin
      + ((normalized.destination_position[index] - origin) * ratio),
  ));
}

export function audioGainAt(automation, elapsedMs = undefined) {
  const normalized = normalizeAudioGainAutomation(automation);
  if (!normalized) {
    throw new TypeError("Invalid audio gain automation");
  }
  const elapsed = elapsedMs ?? normalized.elapsed_ms;
  const ratio = audioMotionProgress(
    elapsed / normalized.duration_ms,
    normalized.easing,
  );
  return normalized.origin_gain
    + ((normalized.destination_gain - normalized.origin_gain) * ratio);
}

export function toWebAudioPosition(position) {
  const [x, y, z] = position;
  // Web Audio's default listener faces -Z with +Y up.
  return Object.freeze([x, z, -y]);
}

export function setAudioSpatializerPosition(panner, context, position) {
  const [x, y, z] = toWebAudioPosition(position);
  if (
    panner.positionX?.setValueAtTime
    && panner.positionY?.setValueAtTime
    && panner.positionZ?.setValueAtTime
  ) {
    panner.positionX.setValueAtTime(x, context.currentTime);
    panner.positionY.setValueAtTime(y, context.currentTime);
    panner.positionZ.setValueAtTime(z, context.currentTime);
  } else if (typeof panner.setPosition === "function") {
    panner.setPosition(x, y, z);
  } else if (panner.pan) {
    panner.pan.setValueAtTime?.(panFromPosition(position), context.currentTime);
    if (!panner.pan.setValueAtTime) {
      panner.pan.value = panFromPosition(position);
    }
  }
}

export function createAudioSpatializer(context, { position, pan = 0 } = {}) {
  const normalizedPosition = normalizeAudioPosition(position);
  if (normalizedPosition === null) {
    return null;
  }
  if (
    normalizedPosition
    && typeof context.createPanner === "function"
  ) {
    try {
      const panner = context.createPanner();
      panner.panningModel = "HRTF";
      // Direction and distance gain are deliberately separate: the shared
      // mixer applies the validated protocol curve exactly once.
      panner.distanceModel = "inverse";
      panner.refDistance = TABLE_RADIUS;
      panner.maxDistance = MAX_AUDIO_POSITION;
      panner.rolloffFactor = 0;
      setAudioSpatializerPosition(panner, context, normalizedPosition);
      return panner;
    } catch {
      // A partial Web Audio implementation still gets the stereo fallback.
    }
  }
  if (typeof context.createStereoPanner === "function") {
    try {
      const panner = context.createStereoPanner();
      const fallbackPan = normalizedPosition
        ? panFromPosition(normalizedPosition)
        : Math.max(-1, Math.min(1, Number.isFinite(Number(pan)) ? Number(pan) : 0));
      panner.pan.value = fallbackPan;
      return panner;
    } catch {
      // Direct connection is the final capability fallback.
    }
  }
  return undefined;
}

File: client/spatial_audio.py
"""Shared spatial position and deterministic distance-gain policy."""

from __future__ import annotations

import math
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any

MAX_AUDIO_POSITION = 1000.0
ATTENUATION_PRECISION = 6
MAX_AUDIO_DISTANCE = (
    math.ceil(math.sqrt(3.0) * MAX_AUDIO_POSITION * 10**ATTENUATION_PRECISION)
    / 10**ATTENUATION_PRECISION
)
MAX_AUDIO_ROLLOFF = 16.0
MAX_AUDIO_AUTOMATION_MS = 3_600_000
MAX_AUDIO_SEQUENCE_SEGMENTS = 32
TABLE_RADIUS = 2.0
AUDIO_ATTENUATION_MODELS = frozenset({"none", "linear", "inverse", "exponential"})
AUDIO_AUTOMATION_EASINGS = frozenset(
    {"linear", "ease-in", "ease-out", "ease-in-out"}
)

AudioPosition = tuple[float, float, float]


@dataclass(frozen=True)
class DistanceAttenuation:
    """Validated client-side representation of the protocol object."""

    model: str
    reference_distance: float | None = None
    max_distance: float | None = None
    rolloff_factor: float | None = None
    min_gain: float | None = None
    max_gain: float | None = None


@dataclass(frozen=True)
class AudioMotion:
    """Validated, resumable position automation from the audio protocol."""

    origin_position: AudioPosition
    destination_position: AudioPosition
    duration_ms: int
    elapsed_ms: int
    easing: str


@dataclass(frozen=True)
class AudioGainAutomation:
    """Validated, resumable source-gain automation."""

    origin_gain: float
    destination_gain: float
    duration_ms: int
    elapsed_ms: int
    easing: str


@dataclass(frozen=True)
class AudioSequenceSegment:
    """Validated finite SFX segment with a decoded-duration onset."""

    asset: str
    position: AudioPosition | None
    destination_position: AudioPosition | None
    attenuation: DistanceAttenuation | None
    gain: float
    easing: str
    next_start_ratio: float


def normalize_audio_gain(value: Any) -> float:
    """Return a finite source gain in the closed unit interval."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError("Audio source gain must be a number")
    gain = float(value)
    if not math.isfinite(gain) or not 0.0 <= gain <= 1.0:
        raise ValueError("Audio source gain is out of range")
    return gain


def normalize_audio_sequence_segments(value: Any) -> tuple[AudioSequenceSegment, ...]:
    """Parse the exact finite-sequence schema without partial playback."""
    if value is None:
        return ()
    if isinstance(value, (str, bytes)) or not isinstance(value, Sequence):
        raise ValueError("Audio sequence segments must be a sequence")
    if not 0 < len(value) <= MAX_AUDIO_SEQUENCE_SEGMENTS:
        raise ValueError("Audio sequence segment count is out of range")
    expected = {
        "asset",
        "position",
        "destination_position",
        "attenuation",
        "gain",
        "easing",
        "next_start_ratio",
    }
    normalized: list[AudioSequenceSegment] = []
    for item in value:
        if isinstance(item, AudioSequenceSegment):
            normalized.append(item)
            continue
        if not isinstance(item, Mapping) or set(item) != expected:
            raise ValueError("Audio sequence segment fields are invalid")
        if not isinstance(item["asset"], str) or not item["asset"]:
            raise ValueError("Audio sequence asset must be a non-empty string")
        asset = item["asset"]
        position = normalize_audio_position(item["position"])
        destination = normalize_audio_position(item["destination_position"])
        attenuation = normalize_distance_attenuation(item["attenuation"])
        gain = normalize_audio_gain(item["gain"])
        next_start_ratio = normalize_audio_gain(item["next_start_ratio"])
        easing = str(item["easing"])
        if attenuation is not None and position is None:
            raise ValueError("Sequence attenuation requires a spatial position")
        if destination is not None and position is None:
            raise ValueError("Sequence motion requires an origin position")
        if easing not in AUDIO_AUTOMATION_EASINGS:
            raise ValueError("Unknown audio sequence easing")
        if destination is None and easing != "linear":
            raise ValueError("Static audio sequence segments require linear easing")
        normalized.append(
            AudioSequenceSegment(
                asset=asset,
                position=position,
                destination_position=destination,
                attenuation=attenuation,
                gain=gain,
                easing=easing,
                next_start_ratio=next_start_ratio,
            )
        )
    return tuple(normalized)


def normalize_audio_position(value: Any) -> AudioPosition | None:
    """Return a finite listener-relative point, or raise for malformed input."""
    if value is None:
        return None
    if isinstance(value, (str, bytes)) or not isinstance(value, Sequence):
        raise ValueError("Audio position must contain three numbers")
    coordinates = list(value)
    if len(coordinates) != 3 or any(
        isinstance(component, bool) or not isinstance(component, (int, float))
        for component in coordinates
    ):
        raise ValueError("Audio position must contain three numbers")
    position = tuple(float(component) for component in coordinates)
    if not all(
        math.isfinite(component) and abs(component) <= MAX_AUDIO_POSITION
        for component in position
    ):
        raise ValueError("Audio position coordinates must be finite and in range")
    return position


def proportional_list_pan(index: int, count: int) -> float:
    """Map a zero-based list index across the full left-to-right field."""
    if count <= 1:
        return 0.0
    bounded_index = max(0, min(count - 1, index))
    return (bounded_index / (count - 1) * 2.0) - 1.0


def frontal_position_for_pan(
    pan: float,
    radius: float = TABLE_RADIUS,
) -> AudioPosition:
    """Place a normalized pan on the listener's forward-facing semicircle."""
    bounded_pan = max(-1.0, min(1.0, float(pan)))
    x = bounded_pan * radius
    y = math.sqrt(max(0.0, radius * radius - x * x))
    return (x, y, 0.0)


def normalize_distance_attenuation(value: Any) -> DistanceAttenuation | None:
    """Parse the exact version-3 attenuation schema without client defaults."""
    if value is None:
        return None
    if isinstance(value, DistanceAttenuation):
        fields = {
            "model": value.model,
            "reference_distance": value.reference_distance,
            "max_distance": value.max_distance,
            "rolloff_factor": value.rolloff_factor,
            "min_gain": value.min_gain,
            "max_gain": value.max_gain,
        }
        parameters = tuple(fields[name] for name in fields if name != "model")
        value = (
            {"model": "none"}
            if value.model == "none" and all(item is None for item in parameters)
            else fields
        )
    if not isinstance(value, Mapping):
        raise ValueError("Audio attenuation must be an object")
    fields = dict(value)
    model = fields.get("model")
    if model not in AUDIO_ATTENUATION_MODELS:
        raise ValueError("Unknown audio attenuation model")
    expected = (
        {"model"}
        if model == "none"
        else {
            "model",
            "reference_distance",
            "max_distance",
            "rolloff_factor",
            "min_gain",
            "max_gain",
        }
    )
    if set(fields) != expected:
        raise ValueError("Audio attenuation has missing or unknown fields")
    if model == "none":
        return DistanceAttenuation(model="none")

    values = [
        fields["reference_distance"],
        fields["max_distance"],
        fields["rolloff_factor"],
        fields["min_gain"],
        fields["max_gain"],
    ]
    if any(
        isinstance(item, bool)
        or not isinstance(item, (int, float))
        or not math.isfinite(float(item))
        for item in values
    ):
        raise ValueError("Audio attenuation parameters must be finite numbers")
    reference_distance, max_distance, rolloff_factor, min_gain, max_gain = map(
        float, values
    )
    if not 0.0 < reference_distance < max_distance <= MAX_AUDIO_DISTANCE:
        raise ValueError("Audio attenuation distances are out of range")
    maximum_rolloff = 1.0 if model == "linear" else MAX_AUDIO_ROLLOFF
    if not 0.0 < rolloff_factor <= maximum_rolloff:
        raise ValueError("Audio attenuation rolloff is out of range")
    if not 0.0 <= min_gain <= max_gain <= 1.0:
        raise ValueError("Audio attenuation gains are out of range")
    return DistanceAttenuation(
        model=model,
        reference_distance=reference_distance,
        max_distance=max_distance,
        rolloff_factor=rolloff_factor,
        min_gain=min_gain,
        max_gain=max_gain,
    )


def normalize_audio_motion(value: Any) -> AudioMotion | None:
    """Parse the exact version-3 position-automation schema."""
    if value is None:
        return None
    if isinstance(value, AudioMotion):
        value = {
            "origin_position": value.origin_position,
            "destination_position": value.destination_position,
            "duration_ms": value.duration_ms,
            "elapsed_ms": value.elapsed_ms,
            "easing": value.easing,
        }
    if not isinstance(value, Mapping):
        raise ValueError("Audio motion must be an object")
    fields = dict(value)
    expected = {
        "origin_position",
        "destination_position",
        "duration_ms",
        "elapsed_ms",
        "easing",
    }
    if set(fields) != expected:
        raise ValueError("Audio motion has missing or unknown fields")
    origin = normalize_audio_position(fields["origin_position"])
    destination = normalize_audio_position(fields["destination_position"])
    if origin is None or destination is None:
        raise ValueError("Audio motion requires origin and destination positions")
    duration_ms = fields["duration_ms"]
    elapsed_ms = fields["elapsed_ms"]
    if (
        isinstance(duration_ms, bool)
        or not isinstance(duration_ms, int)
        or not 0 < duration_ms <= MAX_AUDIO_AUTOMATION_MS
    ):
        raise ValueError("Audio motion duration is out of range")
    if (
        isinstance(elapsed_ms, bool)
        or not isinstance(elapsed_ms, int)
        or not 0 <= elapsed_ms <= duration_ms
    ):
        raise ValueError("Audio motion elapsed time is out of range")
    easing = fields["easing"]
    if easing not in AUDIO_AUTOMATION_EASINGS:
        raise ValueError("Unknown audio motion easing")
    return AudioMotion(origin, destination, duration_ms, elapsed_ms, easing)


def normalize_audio_gain_automation(value: Any) -> AudioGainAutomation | None:
    """Parse the exact version-3 source-gain automation schema."""
    if value is None:
        return None
    if isinstance(value, AudioGainAutomation):
        value = {
            "origin_gain": value.origin_gain,
            "destination_gain": value.destination_gain,
            "duration_ms": value.duration_ms,
            "elapsed_ms": value.elapsed_ms,
            "easing": value.easing,
        }
    if not isinstance(value, Mapping):
        raise ValueError("Audio gain automation must be an object")
    fields = dict(value)
    expected = {
        "origin_gain",
        "destination_gain",
        "duration_ms",
        "elapsed_ms",
        "easing",
    }
    if set(fields) != expected:
        raise ValueError("Audio gain automation has missing or unknown fields")
    origin = normalize_audio_gain(fields["origin_gain"])
    destination = normalize_audio_gain(fields["destination_gain"])
    duration_ms = fields["duration_ms"]
    elapsed_ms = fields["elapsed_ms"]
    if (
        isinstance(duration_ms, bool)
        or not isinstance(duration_ms, int)
        or not 0 < duration_ms <= MAX_AUDIO_AUTOMATION_MS
    ):
        raise ValueError("Audio gain duration is out of range")
    if (
        isinstance(elapsed_ms, bool)
        or not isinstance(elapsed_ms, int)
        or not 0 <= elapsed_ms <= duration_ms
    ):
        raise ValueError("Audio gain elapsed time is out of range")
    easing = fields["easing"]
    if easing not in AUDIO_AUTOMATION_EASINGS:
        raise ValueError("Unknown audio gain easing")
    return AudioGainAutomation(origin, destination, duration_ms, elapsed_ms, easing)


def audio_motion_progress(progress: float, easing: str) -> float:
    """Evaluate the protocol's deterministic quadratic easing functions."""
    bounded = max(0.0, min(1.0, float(progress)))
    if easing == "linear":
        return bounded
    if easing == "ease-in":
        return bounded * bounded
    if easing == "ease-out":
        return 1.0 - (1.0 - bounded) ** 2
    if easing == "ease-in-out":
        return (
            2.0 * bounded * bounded
            if bounded < 0.5
            else 1.0 - ((-2.0 * bounded + 2.0) ** 2) / 2.0
        )
    raise ValueError("Unknown audio motion easing")


def audio_motion_position(
    motion: AudioMotion | Mapping[str, Any],
    elapsed_ms: float | None = None,
) -> AudioPosition:
    """Interpolate one source position at an absolute trajectory time."""
    normalized = normalize_audio_motion(motion)
    if normalized is None:
        raise ValueError("Audio motion is required")
    elapsed = normalized.elapsed_ms if elapsed_ms is None else float(elapsed_ms)
    ratio = audio_motion_progress(elapsed / normalized.duration_ms, normalized.easing)
    return tuple(
        origin + ((destination - origin) * ratio)
        for origin, destination in zip(
            normalized.origin_position,
            normalized.destination_position,
            strict=True,
        )
    )


def audio_gain_at(
    automation: AudioGainAutomation | Mapping[str, Any],
    elapsed_ms: float | None = None,
) -> float:
    """Interpolate one source gain at an absolute automation time."""
    normalized = normalize_audio_gain_automation(automation)
    if normalized is None:
        raise ValueError("Audio gain automation is required")
    elapsed = normalized.elapsed_ms if elapsed_ms is None else float(elapsed_ms)
    ratio = audio_motion_progress(elapsed / normalized.duration_ms, normalized.easing)
    return normalized.origin_gain + (
        (normalized.destination_gain - normalized.origin_gain) * ratio
    )


def distance_attenuation_gain(
    position: AudioPosition | None,
    attenuation: DistanceAttenuation | Mapping[str, Any] | None,
) -> float:
    """Calculate the canonical distance gain used by every first-party client."""
    normalized = normalize_distance_attenuation(attenuation)
    if normalized is None or normalized.model == "none":
        return 1.0
    if position is None:
        raise ValueError("Audio attenuation requires a spatial position")
    distance = math.sqrt(sum(coordinate * coordinate for coordinate in position))
    clamped_distance = min(
        max(distance, normalized.reference_distance), normalized.max_distance
    )
    if normalized.model == "linear":
        gain = 1.0 - normalized.rolloff_factor * (
            (clamped_distance - normalized.reference_distance)
            / (normalized.max_distance - normalized.reference_distance)
        )
    elif normalized.model == "inverse":
        gain = normalized.reference_distance / (
            normalized.reference_distance
            + normalized.rolloff_factor
            * (clamped_distance - normalized.reference_distance)
        )
    else:
        gain = (
            clamped_distance / normalized.reference_distance
        ) ** -normalized.rolloff_factor
    return max(normalized.min_gain, min(normalized.max_gain, gain))
